src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,219 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import asyncio
+import logging
+import sys
+from collections.abc import Coroutine
+from typing import Any, Callable
+
+import inmanta.deploy.scheduler
+from inmanta.agent import config as agent_config
+from inmanta.data.model import ResourceIdStr
+from inmanta.deploy.work import TaskPriority
+from inmanta.util import CronSchedule, ScheduledTask, Scheduler
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ResourceTimer:
+    def __init__(self, resource: ResourceIdStr):
+        """
+        :param resource: The resource for which to ensure periodic repairs.
+        """
+        self.resource: ResourceIdStr = resource
+        self.time_to_next_deploy = sys.maxsize
+
+        self.repair_handle: asyncio.Task[None] | None = None
+        self.next_schedule_handle: asyncio.TimerHandle | None = None
+
+    def ensure_timer(
+        self,
+        periodic_deploy_interval: int | None,
+        periodic_repair_interval: int | None,
+        is_dirty: bool,
+        action_function: Callable[..., Coroutine[Any, Any, None]],
+    ) -> None:
+        """
+        Make sure the underlying resource is marked for execution
+
+        :param periodic_deploy_interval: Per-resource deploy interval, assumed to be a positive int or None
+        :param periodic_repair_interval: Per-resource repair interval, assumed to be a positive int or None
+        :param is_dirty: Last known state of the resource at the timer install request time.
+        :param action_function: The function to execute
+        """
+
+        next_execute_time: int
+
+        if periodic_repair_interval is None:
+            periodic_repair_interval = sys.maxsize
+        if periodic_deploy_interval is None:
+            periodic_deploy_interval = sys.maxsize
+
+        if is_dirty:
+            next_execute_time = min(periodic_deploy_interval, periodic_repair_interval)
+        else:
+            next_execute_time = periodic_repair_interval
+
+        if next_execute_time == sys.maxsize:
+            return
+
+        # We have to schedule the next deploy sooner than previously though:
+        # Cancel previous schedule and re-schedule
+        if next_execute_time < self.time_to_next_deploy:
+            self.cancel_next_schedule()
```

Comment:
Any time difference leads to a re-schedule, I think? 

(i.e. if we do a deploy, we move all timers to a later place)

Reply:
What about if we go from dirty to clean by receiving an event? 

The deploy time hasn't fired, but it will and it shouldn't.

Reply:
I think that all transitions from / to dirty happen either when a new version comes in or when a resource finishes deploying. But I do think that overall it would be a lot simpler to reason on if whatever comes from the scheduler is considered the authority, relative to now, replacing what was previously there (even if there should never be anything there in practice).